### ISSUE TITLE
When in Japanese input, 'enter' should not immediately send the message

### ISSUE THREAD

**Issue by dvcrn** [ajbura/cinny#2103]
### Describe the bug

In Japanese IME, you typically type text, then hit enter to "commit" the correct Kanji. For example, if I type テスト then it will be underlined until I hit 'enter', in effect confirming that I've finished writing and picked the correct Kanji/characters

Cinny binds to enter, so even though I am not done typing and only committed a word, as soon as I hit enter, Cinny sends the message 

It's a very common problem when using `onkeyup` as listener, eg: https://stackoverflow.com/questions/46812908/enter-keyup-event-in-japanese-input

### Platform and versions

```shell
macOS Desktop 4.2.3, macOS 15.2
```

**Comment by kfiven** [comment]
There's an option to change the behavior of Enter in settings, have you tried that?

**Comment by dvcrn** [comment]
Yes that works, but it's not the expected behavior. Hitting enter when in Japanese (or other asian) IME should not trigger sending of a message, since it will send an incomplete sentence/word

**Comment by programmablereya** [comment]
I haven't encountered this yet, but I can definitely speak to this being undesired, buggy behavior.

Between [compositionstart](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event) and [compositionend](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionend_event) events - i.e., when [KeyboardEvent.isComposing](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/isComposing) is true, the keydown and keyup events should be ignored by the editor. This is because at that time, the IME is composing the user's next character(s). While in an IME, the browser ignores all keystrokes, even ones with modifiers; Ctrl+T does not open a new tab, Ctrl+W does not close the current tab, Tab does not navigate the focus to the next element, etc. The only keystrokes that are expected to still function are system-wide keystrokes like Alt+F4.

The keydown events are still delivered to the browser (and thus to your code), but they should be ignored; they are fired primarily for you to be able to do things like have your code respond to the length of the partially-composed text, not to be treated as input directly to you. [See the guidance here](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event#keydown_events_with_ime) for another source on how to handle keydown events with IMEs.

You can experience an IME and see how it behaves in most programs by installing an input language with one (e.g., Japanese) and switching to it in your input selector. Or, if you're using GNOME, you can press Ctrl+Shift+E to enter an IME for inputting special characters (particularly emoji, but it lists all Unicode characters).

**Comment by programmablereya** [comment]
![Safari bug image, showing an example where Safari sends compositionend-keydown-keyup and Firefox and Chrome send keydown-compositionend-keyup](https://bug-165004-attachments.webkit.org/attachment.cgi?id=470455)

Okay - it looks like the issue is _specifically_ that [Safari sends compositionend before sending the Enter key**down** event](https://bugs.webkit.org/show_bug.cgi?id=165004), whereas other browsers send it between keydown and keyup. isComposing is just a helper property to avoid having to capture compositionstart and compositionend yourself, so it will also be affected by this.

And it's not guaranteed that the next keydown/keyup after compositionend is the keypress that ended composition, even in Safari; there are ways to end composition that don't require a keypress! (e.g., clicking a suggestion with the mouse or a touch input)

However, it _does_ look like Safari respects the idea that `keyCode` must be 229 during composition (but not that `key` must be `Unidentified`...). Which should let us intercept this.

I'll adapt my in-progress patch to make use of this information.

**Comment by programmablereya** [comment]
@dvcrn - could you build the PR version and test that it does, in fact, work around the problem here?

**Comment by dvcrn** [comment]
Tested your PR and it fixes the issues, however there are a few build errors, I'll comment in the PR

**Comment by kfiven** [comment]
The PR by @programmablereya is deployed on https://2175--pr-cinny.netlify.app/ 

If people can test this, it would be helpful.

**Comment by chrisdedman** [comment]
> The PR by [@programmablereya](https://github.com/programmablereya) is deployed on https://2175--pr-cinny.netlify.app/
> 
> If people can test this, it would be helpful.

Tested on Firefox browser. It works as expected.
